tools/doc-generator/doc-generator.go

```diff
@@ -23,99 +60,129 @@ func main() {
+		checkError(err)
+
+		writeToFile(metricByTypes)
 	} else {
-		panic(recorder.Code)
+		panic(fmt.Errorf("got HTTP status code of %d from /metric", recorder.Code))
 	}
 }
 
-func parseVirtMetrics(r io.Reader, contains []string) map[string]string {
-	metrics := map[string]string{}
-	scan := bufio.NewScanner(r)
-	for scan.Scan() {
-		for _, contain := range contains {
-			help := scan.Text()
-			if strings.Contains(help, contain) && strings.HasPrefix(help, "# HELP ") {
-				name := strings.Split(help, " ")[2]
-
-				metrics[name] = help
+func writeToFile(metricByTypes []metricByType) {
+	newFile, err := os.Create("newmetrics.md")
+	checkError(err)
+	defer newFile.Close()
 
-			}
-		}
+	fmt.Fprint(newFile, opening)
+	for _, metrics := range metricByTypes {
+		metrics.writeToFile(newFile)
 	}
-	if scan.Err() != nil {
-		panic(fmt.Errorf("Failed to parse metrics from prometheus endpoint, %v", scan.Err()))
-	}
-	return metrics
+	fmt.Fprint(newFile, footer)
+
 }
 
-func writeToFile(metrics map[string]string) {
-	old, err := os.Open("metrics.md")
-	checkError(err)
-	defer old.Close()
+type metric struct {
+	name        string
+	description string
+}
 
-	new, err := os.Create("newmetrics.md")
-	checkError(err)
-	defer new.Close()
+func (m metric) writeToFile(newFile io.WriteCloser) {
+	fmt.Fprintln(newFile, "###", m.name)
+	fmt.Fprintln(newFile, m.description)
+	fmt.Fprintln(newFile)
+}
 
-	var write string
+type metricList []metric
 
-	format := "###%s"
+func (m metricList) Len() int {
+	return len(m)
+}
 
-	buf := bufio.NewWriter(new)
-	scan := bufio.NewScanner(old)
-	for scan.Scan() {
-		line := scan.Text()
-		if write != "" {
-			if strings.Contains(line, "HELP") {
-				if line != write {
-					buf.WriteString(write)
-					line = ""
-				}
+func (m metricList) Less(i, j int) bool {
+	return m[i].name < m[j].name
+}
 
-			} else {
-				buf.WriteString(write)
-				buf.WriteString("\n")
-			}
-			write = ""
+func (m metricList) Swap(i, j int) {
+	m[i], m[j] = m[j], m[i]
+}
 
-		}
+func (m *metricList) add(line string) {
+	split := strings.Split(line, " ")
+	name := split[2]
+	split[3] = strings.Title(split[3])
+	description := strings.Join(split[3:], " ")
+	*m = append(*m, metric{name: name, description: description})
+}
 
-		buf.WriteString(line + "\n")
-		for k, v := range metrics {
-			if strings.Contains(line, k) {
-				write = fmt.Sprintf(format, v)
-				delete(metrics, k)
-				break
-			}
-		}
-	}
-	if scan.Err() != nil {
-		panic(fmt.Errorf("Failed to update metric doc, %v", scan.Err()))
+func (m metricList) writeToFile(newFile io.WriteCloser) {
+	sort.Sort(m)
+
+	for _, met := range m {
+		met.writeToFile(newFile)
 	}
+}
+
+type metricByType struct {
+	filter  string
+	heading string
+	metrics metricList
+}
+
+func (m metricByType) writeToFile(newFile io.WriteCloser) {
+	fmt.Fprintf(newFile, "## %s Metrics\n", m.heading)
+	m.metrics.writeToFile(newFile)
+}
 
-	if len(metrics) > 0 {
-		buf.WriteString("\n # Other Metrics \n")
+var (
```

Comment:
We need all kubevirt metrics that we get from /metrics. Not only VMI and Virt. So this is not needed.

Reply:
We just need to sort the by name